[mlir][ArmSME] Switch to an attribute-based tile allocation scheme

MacDue opened the pull request:
This reworks the ArmSME dialect to use attributes for tile allocation. This has a number of advantages and corrects some issues with the previous approach:

 * Tile allocation can now be done ASAP (i.e. immediately after `-convert-vector-to-arm-sme`)
 * SSA form for control flow is now supported (e.g.`scf.for` loops that yield tiles)
 * ArmSME ops can be converted to intrinsics very late (i.e. after lowering to control flow)
 * Tests are simplified by removing constants and casts
 * Avoids correctness issues with representing LLVM `immargs` as MLIR values
    - The tile ID on the SME intrinsics is an `immarg` (so is required to be a compile-time constant), `immargs` should be mapped to MLIR attributes (this is already the case for intrinsics in the LLVM dialect)
    - Using MLIR values for `immargs` can lead to invalid LLVM IR being generated (and passes such as -cse making incorrect optimizations)

As part of this patch we bid farewell to the following operations:

```mlir
arm_sme.get_tile_id : i32
arm_sme.cast_tile_to_vector : i32 to vector<[4]x[4]xi32>
arm_sme.cast_vector_to_tile : vector<[4]x[4]xi32> to i32
```

These are now replaced with:
```mlir
// Allocates a new tile with (indeterminate) state:
arm_sme.get_tile : vector<[4]x[4]xi32>
// A placeholder operation for lowering ArmSME ops to intrinsics:
arm_sme.materialize_ssa_tile : vector<[4]x[4]xi32>
```

The new tile allocation works by operations implementing the `ArmSMETileOpInterface`. This interface says that an operation needs to be assigned a tile ID, and may conditionally allocate a new SME tile.

Operations allocate a new tile by implementing...
```c++
std::optional<arm_sme::ArmSMETileType> getAllocatedTileType()
```
...and returning what type of tile the op allocates (ZAB, ZAH, etc).

Operations that don't allocate a tile return `std::nullopt` (which is the default behaviour).

Currently the following ops are defined as allocating:
```mlir
arm_sme.get_tile
arm_sme.zero
arm_sme.tile_load
arm_sme.outerproduct // (if no accumulator is specified)
```

Allocating operations become the roots for the tile allocation pass, which currently just (naively) assigns all transitive uses of a root operation the same tile ID. However, this is enough to handle current use cases.

Once tile IDs have been allocated subsequent rewrites can forward the tile IDs to any newly created operations.

MacDue committed:
[mlir][ArmSME] Switch to an attribute-based tile allocation scheme

This reworks the ArmSME dialect to use attributes for tile allocation.
This has a number of advantages and corrects some issues with the
previous approach:

 * Tile allocation can now be done ASAP (i.e. immediately after
   `-convert-vector-to-arm-sme`)
 * SSA form for control flow is now supported (e.g.`scf.for` loops that
   yeild tiles)
 * ArmSME ops can be converted to intrinsics very late (i.e. after
   lowering to control flow)
 * Tests are simplified by removing constants and casts
 * Avoids correctness issues with representing LLVM `immargs` as MLIR
   values
    - The tile ID on the SME intrinsics is an `immarg` (so is required
      to be a compile-time constant), `immargs` should be mapped to
      MLIR attributes (this is already the case for intrinsics in the
      LLVM dialect)
    - Using MLIR values for `immargs` can lead to invalid LLVM IR being
      generated (and passes such as -cse making incorrect optimizations)

As part of this patch we bid farewell to the following operations:

```mlir
arm_sme.get_tile_id : i32
arm_sme.cast_tile_to_vector : i32 to vector<[4]x[4]xi32>
arm_sme.cast_vector_to_tile : vector<[4]x[4]xi32> to i32
```

These are now replaced with:
```mlir
// Allocates a new tile with (indeterminate) state:
arm_sme.get_tile : vector<[4]x[4]xi32>
// A placeholder operation for lowering ArmSME ops to intrinsics:
arm_sme.materialize_ssa_tile : vector<[4]x[4]xi32>
```

The new tile allocation works by operations implementing the
`ArmSMETileOpInterface`. This interface says that an operation needs
to be assigned a tile ID, and may conditionally allocate a new SME tile.

Operations allocate a new tile by implementing...
```c++
std::optional<arm_sme::ArmSMETileType> getAllocatedTileType()
```
...and returning what type of tile the op allocates (ZAB, ZAH, etc).

Operations that don't allocate a tile return `std::nullopt` (which
is the default behaviour).

Currently the following ops are defined as allocating:
```mlir
arm_sme.get_tile
arm_sme.zero
arm_sme.tile_load
arm_sme.outerproduct // (if no accumulator is specified)
```

Allocating operations become the roots for the tile allocation pass,
which currently just (naively) assigns all transitive uses of a root
operation the same tile ID. However, this is enough to handle current
use cases.

Once tile IDs have been allocated subsequent rewrites can forward the
tile IDs to any newly operations.

banach-space commented on mlir/tools/mlir-query/mlir-query.cpp:
```diff
@@ -21,9 +21,9 @@
-using HasOpAttrName = detail::AttrOpMatcher(StringRef);
-using HasOpName = detail::NameOpMatcher(StringRef);
-using IsConstantOp = detail::constant_op_matcher();
+using HasOpAttrName = mlir::detail::AttrOpMatcher(StringRef);
+using HasOpName = mlir::detail::NameOpMatcher(StringRef);
+using IsConstantOp = mlir::detail::constant_op_matcher();
 
```
Unrelated changes?

MacDue replied:
I'm not sure... For some reason these are reported as ambiguous if I don't prefix them with `mlir::` now, not worked out why yet.

banach-space commented on mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEOps.td:
```diff
@@ -21,6 +21,99 @@ include "mlir/Interfaces/SideEffectInterfaces.td"
+    | `vector<[2]x[2]xi64>` or `vector<[2]x[2]xf64>`                          | 0 to 7 (inclusive)  |
+    | `vector<[1]x[1]xi128>`                                                  | 0 to 15 (inclusive) |
+
+    Operations that allocate a new tiles (such as arm_sme.get_tile), are used as
```
```suggestion
    Operations that allocate a new tile (such as arm_sme.get_tile), are used as
```

banach-space commented on mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEOps.td:
```diff
@@ -21,6 +21,99 @@ include "mlir/Interfaces/SideEffectInterfaces.td"
+def ArmSMETileOpInterface : OpInterface<"ArmSMETileOpInterface"> {
+  let description = [{
+    An interface for operations that use or allocate Arm SME tiles. These
+    operations need to be assigned a tile ID an i32 attribute, which specifies
```
```suggestion
    operations need to be assigned a tile ID, an i32 attribute, which specifies
```

banach-space commented on mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEOps.td:
```diff
@@ -21,6 +21,99 @@ include "mlir/Interfaces/SideEffectInterfaces.td"
+      }]
+    >,
+    InterfaceMethod<
+      "Returns the (possibly null) tile ID assigned to this operation.",
```
Does `null` correspond to  "tile allocation hasn't been run yet"?

MacDue replied:
Yep, just that no tile ID has been allocated yet.

banach-space commented on mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEOps.td:
```diff
@@ -21,6 +21,99 @@ include "mlir/Interfaces/SideEffectInterfaces.td"
+      }]
+    >,
+    InterfaceMethod<
+      "The type of tile this operation allocates (or none)",
```
What does "none" correspond to?

MacDue replied:
It means this operation does not allocate a tile.

banach-space commented on mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEOps.td:
```diff
@@ -21,6 +21,99 @@ include "mlir/Interfaces/SideEffectInterfaces.td"
+      return op;
+    }
+
+    // A helper to replace this operation and forward any tile ID.
```
nice! 
```suggestion
    // A helper to replace this operation and forward its tile ID (if present).
```
?

banach-space commented on mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEOps.td:
```diff
@@ -162,125 +222,67 @@ def ArmSME_CombiningKindAttr : EnumAttr<ArmSME_Dialect, CombiningKind,
-    normally be used when lowering operations that return "virtual tile" vector
-    types to model the output. This is required to preserve dataflow as SME
-    intrinsics have no return values.
+    Allocates a new SME "virtual tile" within a function. The contents of the
+    tile returned from this operation undefined.
```
Why "within a function"?

Also:
```suggestion
    Allocates a new SME "virtual tile" within a function. The contents of the
    tile returned from this operation are undefined.
```

MacDue replied:
Tiles are only allocated with respect to a function right now, but they don't necessarily need to be, MLIR is pretty high-level in SCF, so tiles could be allocated in scopes (like in C).. which could be a something that's useful.

banach-space commented on mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEOps.td:
```diff
@@ -162,125 +222,67 @@ def ArmSME_CombiningKindAttr : EnumAttr<ArmSME_Dialect, CombiningKind,
-    The scope of a tile id is a function and cannot be passed or returned from
-    functions.
+    A placeholder to preserve dataflow while lowering to SME intrinsics (which
+    do not take or return tile values). This operation is intended to be DCE'd
```
```suggestion
    do not take or return SME virtual tile values). This operation is intended to be DCE'd
```

banach-space commented on mlir/lib/Dialect/ArmSME/Transforms/TileAllocation.cpp:
```diff
@@ -107,73 +109,85 @@ enum class TileMask : unsigned {
+    getForwardSlice(tileOp.getOperation(), &dependantOps);
+
+    // Set all operations to use the same tile ID.
+    // This is a navie tile allocation scheme, but works for common cases.
```
```suggestion
    // This is a naive tile allocation scheme, but works for common cases.
```
Why is it naive? Could you elaborate in the comment? My understanding is that all ops that depend on this tile _have to_ use the same `tile_id`. Is that correct?

MacDue replied:
I think that's correct. I more mean that to correctly allocate tiles in all cases would require more inspection of the program, and inserting new ops like tile moves. 

For example you could end up with a case like:
```
%tile = scf.if %some_cond -> vector<[4]x[4]xi32> {
   scf.yield %tileA {tile_id = 0} : vector<[4]x[4]xi32>
} else {
   scf.yield %tileB {tile_id = 1} : vector<[4]x[4]xi32>
}
```
Which would need to become (something like):
```
%ifResult = arm_sme.get_tile  {tile_id = 2} : vector<[4]x[4]xi32>
%tile = scf.if %some_cond -> vector<[4]x[4]xi32> {
   arm_sme.move_tile %ifResult <- %tileA : vector<[4]x[4]xi32>
   scf.yield %ifResult {tile_id = 2} : vector<[4]x[4]xi32>
} else {
   arm_sme.move_tile %ifResult <- %tileB : vector<[4]x[4]xi32>
   scf.yield %tileB {tile_id = 2} : vector<[4]x[4]xi32>
}

banach-space commented on mlir/lib/Dialect/ArmSME/Transforms/TileAllocation.cpp:
```diff
@@ -107,73 +109,85 @@ enum class TileMask : unsigned {
-  auto masks = getMasks(tileIDOp.getType());
-  for (const auto &it : llvm::enumerate(masks)) {
-    const auto tileMask = it.value();
+/// Allocates a tile to 'tileId' or returns an error if there are no tiles left.
```
[nit] `tileId` feels like an implementation detail that I'd skip in a comment documenting a method. Perhaps "Allocates and returns a tile id. Returns an error if there are no tiles left".

Perhaps also rename - otherwise it clashes with `arm_sme.get_tile`.

banach-space reviewed (commented):
Looks great, thanks! There's quite a bit of churn, but sadly that's unavoidable. 

I've left a few minor comments - nothing major. I'd like to take another look though - this is quite dense 😅 .

MacDue committed:
Fixups

MacDue committed:
Update comments that reference removed operations

MacDue committed:
Add explanation

c-rhodes commented on mlir/test/Integration/Dialect/Linalg/CPU/ArmSME/matmul-transpose-a.mlir:
```diff
@@ -3,10 +3,10 @@
 // RUN:   -enable-arm-streaming="streaming-mode=streaming-locally za-mode=new-za" \
 // RUN:   -convert-vector-to-arm-sme -convert-arm-sme-to-scf \
-// RUN:   -convert-vector-to-scf -cse -arm-sve-legalize-vector-storage \
+// RUN:   -convert-vector-to-scf -allocate-arm-sme-tiles -cse -arm-sve-legalize-vector-storage \
```
move before `-convert-vector-to-scf`?

c-rhodes commented on mlir/test/Integration/Dialect/Linalg/CPU/ArmSME/fill-2d.mlir:
```diff
@@ -4,9 +4,9 @@
 // RUN:   -one-shot-bufferize="bufferize-function-boundaries" \
 // RUN:   -enable-arm-streaming="streaming-mode=streaming-locally za-mode=new-za" \
-// RUN:   -convert-vector-to-arm-sme -convert-arm-sme-to-scf \
+// RUN:   -convert-vector-to-arm-sme -convert-arm-sme-to-scf -allocate-arm-sme-tiles  \
```
you mentioned `-allocate-arm-sme-tiles` can happen immediately after `-convert-vector-to-arm-sme`, is this the case here?

c-rhodes reviewed (commented):
Great work Ben, I've left a few minor questions / comments but otherwise LGTM, really polished

MacDue commented on mlir/lib/Conversion/ArmSMEToSCF/ArmSMEToSCF.cpp:
```diff
@@ -506,6 +496,7 @@ struct TileVectorPrintOpConversion : public OpRewritePattern<vector::PrintOp> {
       rewriter.setInsertionPointToStart(forOp.getBody());
       // Extract the current row from the tile.
       Value rowIndex = forOp.getInductionVar();
+      // FIXME: Forward tile IDs.
```
It's because `vector.print` is not a `ArmSMETileOp`, so unlike the other ops this would require looking at the use-def chain of the tile.

MacDue replied:
As it's only for debug, I didn't think it was crucial :smile:

c-rhodes replied:
ok thanks for the clarification. I thought this was currently used in the integration tests?

MacDue replied:
It's used it some, but not all. The simple workaround is to do `-allocate-arm-sme-tiles` after `-convert-arm-sme-to-scf` (since this lowers it to `ArmSMETileOps`).

MacDue replied:
Done :+1:

MacDue replied:
Another solution is just to move this to `convert-vector-to-arm-sme`, which I think is fine (as we already generate loops for arith ops in there).

MacDue replied:
(That can be a follow up patch though)

c-rhodes replied:
This got me thinking. The tile allocation works by assigning tile IDs for the root operations that implement the `ArmSMETileOpInterface` and allocate tiles. An operation allocates a tile if it implements `getAllocatedTileType`, this includes:
```mlir
arm_sme.get_tile
arm_sme.zero
arm_sme.tile_load
arm_sme.outerproduct // (if no accumulator is specified)
```
the tile ID is then forwarded to any uses of the root that also implement the `ArmSMETileOpInterface`. Is this constraint necessary? If the tile ID could be forwarded to any use (i.e. non-ArmSME ops) could it fix this problem?

An alternative of course is to mirror ops such as `vector.print` in the ArmSME dialect with ops that implement this interface.

MacDue replied:
I think the correct fix here is just to more the vector.print lowering into `-convert-vector-to-arm-sme` as it's a vector op lowering, not an ArmSME op lowering.

MacDue replied:
Which is how other similar tile ops are lowered (e.g. vector.splat, vector.broadcast, etc). Using an interface allows setting/getting the tile IDs via methods (and validating them), it also avoids possible conflicts with other dialects (and prevents the IR from getting very messy.

c-rhodes replied:
> I think the correct fix here is just to more the vector.print lowering into `-convert-vector-to-arm-sme` as it's a vector op lowering, not an ArmSME op lowering (no extra op required).

Ok, I think that's fine for now. As you mentioned, this doesn't have to be part of this patch. We may also have to revisit loop materialization in `-convert-vector-to-arm-sme`. When loads/stores were first implemented loop materialization was done during lowering to intrinsics, `-convert-arm-sme-to-scf` was introduced to break that up such that the intrinsic lowerings were 1-1. It's not quite the same when lowering Vector ops to ArmSME, but there may be value in getting the granularity right, that could necessitate more ops in ArmSME however, so some cost there at least.

MacDue committed:
Fixups

c-rhodes reviewed (approved):
LGTM cheers

MacDue committed:
Clarify FIXME

c-rhodes commented on mlir/lib/Conversion/ArmSMEToLLVM/ArmSMEToLLVM.cpp:
```diff
@@ -32,24 +32,41 @@ using namespace mlir;
+  LogicalResult
+  matchAndRewrite(arm_sme::GetTileOp getTile, OpAdaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    rewriter.replaceOpWithNewOp<arm_sme::MaterializeSSATileOp>(
```
Could `arm_sme.materialize_ssa_tile` be replaced with `llvm.mlir.undef`?

MacDue replied:
No, `llvm.mlir.undef` only suports LLVM types, and there's no LLVM type for SME tiles.

MacDue committed:
Implement more semantically correct findDependantOps()

This function follows uses of a value through control flow. It understands
basic SCF contructs and more generally works on control flow branches.

(the previous slice analysis is very basic and does not understand any
control flow)

MacDue committed:
Check for invalid tile allocations

MacDue committed:
Use named op helpers

banach-space commented on mlir/lib/Dialect/ArmSME/Transforms/TileAllocation.cpp:
```diff
@@ -107,73 +107,151 @@ enum class TileMask : unsigned {
+      findDependantOps(exitValues[idx], dependantOps);
+    }
+  };
+  for (Operation *user : rootValue.getUsers()) {
+    if (dependantOps.contains(user))
+      continue;
+    dependantOps.insert(user);
+    if (auto branchOp = llvm::dyn_cast<cf::BranchOp>(user)) {
+      // (CF) Follow branch.
+      traverseCorrespondingValues(branchOp.getDestOperands(),
+                                  branchOp.getDest()->getArguments());
+    } else if (auto condBranchOp = llvm::dyn_cast<cf::CondBranchOp>(user)) {
+      // (CF) Follow true branch.
+      traverseCorrespondingValues(condBranchOp.getTrueOperands(),
+                                  condBranchOp.getTrueDest()->getArguments());
+      // (CF) Follow false branch.
+      traverseCorrespondingValues(condBranchOp.getFalseOperands(),
+                                  condBranchOp.getFalseDest()->getArguments());
+    } else if (auto loop = llvm::dyn_cast<LoopLikeOpInterface>(user)) {
+      // (SCF) Follow iter_args of (basic) loops (e.g. for loops).
+      traverseCorrespondingValues(loop.getInits(), loop.getRegionIterArgs());
+    } else if (user->hasTrait<OpTrait::ReturnLike>()) {
+      // (SCF) Follow yields of (basic) control flow (e.g. for loops).
+      auto parent = user->getParentOp();
+      // Don't traverse outside a function.
+      if (llvm::isa<FunctionOpInterface>(parent))
+        continue;
+      traverseCorrespondingValues(user->getOperands(), parent->getResults());
+    } else {
+      // Otherwise, assume users of _any_ result are dependant.
+      for (Value result : user->getResults())
+        findDependantOps(result, dependantOps);
+    }
+  }
```
Please avoid nested `if` :)

https://llvm.org/docs/CodingStandards.html#use-early-exits-and-continue-to-simplify-code

MacDue replied:
I'm not really sure how that applies here? My understanding of early exits is if you were doing something like:
```
if (A && B && C) {
  // Do thing
}
return null;
```
It's nicer to rewrite it as:
```
if (!A || !B || !C)
  return null;

// Do thing
```
Which make sense :)

But this is checking a number of independent cases, so seems most natural as if/else?

joker-eph replied:
Right it's not a clear cut here. If you wanted something more structured, there is `TypeSwitch`:

```
TypeSwitch<Operation *>(user)
          .Case<cf::BranchOp>(
              [&](auto brancOp) {
                // (CF) Follow branch.
               traverseCorrespondingValues(branchOp.getDestOperands(),
                                  branchOp.getDest()->getArguments());
           }).Case<...>...
```

banach-space replied:
> I'm not really sure how that applies here? My understanding of early exits is if you were doing something like:

You can also do this:
```cpp
if  (A) {
 // Do thing
 continue;
 }
 
if (B) {
  // Do thing
  continue;
}
```

`TypeSwitch` is obviously much nicer :)

banach-space reviewed (commented):
Mostly nits. I'd like to take another look at tile allocation - just to better understand the limitations. The overall design makes a lot of sense.

joker-eph commented on mlir/lib/Dialect/ArmSME/Transforms/TileAllocation.cpp:
```diff
@@ -107,73 +107,151 @@ enum class TileMask : unsigned {
+    if (failed(tileId))
+      return tileOp.emitError("ran out of SME virtual tiles!");
+
+    func->setAttr(kTilesInUseAttr,
```
Please use consistently `getDiscardableAttr` and `setDiscardableAttr` in newly written code.

You code needs to be resilient to these attribute being "discarded", by the nature of it!

MacDue replied:
Done :+1:

MacDue committed:
Fixups

MacDue committed:
Minor change to reattempt CI

banach-space reviewed (approved):
Makes a lot of sense, LGTM :)

It's been a huge effort - thank you for working on this! This change is a significant improvement the SME support in MLIR 🙏🏻 🙇🏻 .

(please wait for +1 from @joker-eph before landing this)

joker-eph commented:
> (please wait for +1 from @joker-eph before landing this)

I didn't look deeper than the high-level drive-by comments I made before, so please don't wait for me :)

dcaballe reviewed (commented):
Awesome! This is a big improvements! Thanks!
I have just one design comment.

MacDue commented on mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEOps.td:
```diff
@@ -162,125 +230,67 @@ def ArmSME_CombiningKindAttr : EnumAttr<ArmSME_Dialect, CombiningKind,
-    bitwidth of the "virtual tile".
+    A placeholder to preserve dataflow while lowering to SME intrinsics (which
+    do not take or return SME virtual tile values). This operation is intended
+    to be DCE'd once all ArmSME operations have been lowered.
```
This idea is shown in: https://github.com/MacDue/llvm-project/blob/b9064adf0d6a53c61a907fc2e61f51997a651eab/mlir/test/Dialect/ArmSME/canonicalize.mlir (where the `arm_sme.materialize_ssa_tile` op and the unused tile SSA values fold way).
